### PR TITLE
Add hardware console log to IPMI and PowerVM backends

### PR DESCRIPTION
**Pull request by czerw**
Enhancement poo#103903: IPMI and PowerVM backends miss log from system
boot/shutdown or kernel crash. This feature enables direct capture of
ipmi/mkvterm sessions on IPMI and PowerVM backends. It is disabled by
default, managed by variable HARDWARE_CONSOLE_LOG. New log file is
hardware-console-log.txt.

Progress: https://progress.opensuse.org/issues/103903
Needles: none

Verification:
ipmi: http://10.100.12.105/tests/1536/logfile?filename=hardware-console-log.txt
spvm: http://10.100.12.105/tests/1530/logfile?filename=hardware-console-log.txt
hmc: http://10.100.12.105/tests/1537/logfile?filename=hardware-console-log.txt

### CODE REVIEW

File: consoles/localXvnc.pm
Comment (Martchus) [inline consoles/localXvnc.pm:41]: This would be a bit easier to read in my opinion:
```suggestion
    $command = "script -f hardware-console-log.txt -c \"$command\"" if $self->{args}->{log};
```

**Review by Martchus — COMMENTED** [review]
I suppose this should work. The duplication for the different backends isn't nice but I suppose it is ok. `script` is part of `util-linux` so we likely don't have to add it as dependency.

**Comment by czerw** [comment]
>The duplication for the different backends isn't nice but I suppose it is ok. 

Do you mean duplication in variables doc? Yeas, I was not happy as well, but not sure how to do it  better.

File: consoles/localXvnc.pm
Comment (czerw) [reply consoles/localXvnc.pm:41]: Implemented.

**Review by Martchus — APPROVED** [review]

**Comment by Martchus** [comment]
The pipeline error is unrelated:

```
(11/11) Installing: perl-App-cpanminus-1.7043-bp153.1.12.noarch [...........done]
Abort, retry, ignore? [a/r/i/...? shows all options] (a): a
File './repodata/2ec0d1f23dc67a2387797d873e045a03a9b89f0db8c421f59a4b5678fdbc9582-deltainfo.xml.gz' not found on medium 'http://download.opensuse.org/update/leap/15.3/sle/'
Error building the cache:
Warning: Skipping repository 'Update repository with updates from SUSE Linux Enterprise 15' because of the above error.
[repo-sle-update|http://download.opensuse.org/update/leap/15.3/sle/] Valid metadata not found at specified URL
History:
 - ABORT request: Aborting requested by user
 - Can't provide ./repodata/2ec0d1f23dc67a2387797d873e045a03a9b89f0db8c421f59a4b5678fdbc9582-deltainfo.xml.gz
```

In fact, I've just seen this error in a completely different context. Let's see whether a retry helps.

By the way, maybe you can also extend unit tests to cover this case.

File: backend/ipmi.pm
Comment (Martchus) [inline backend/ipmi.pm:100]: Maybe tidy let's you do something like this (which would look less ugly):
```suggestion
    my %sol_args = (
        persistent => $bmwqemu::vars{IPMI_SOL_PERSISTENT_CONSOLE} // 1,
        log => $bmwqemu::vars{HARDWARE_CONSOLE_LOG} // 0,
    );
    my $sol = $testapi::distri->add_console('sol', 'ipmi-xterm', \%sol_args);
```

**Comment by czerw** [comment]
> By the way, maybe you can also extend unit tests to cover this case.

But how? It depends on network and real hardware, to be honest i'm not good at this mysterious perl mock area :)

**Review by okurz — CHANGES_REQUESTED** [review]
IMHO the duplication is acceptable. I would have approved if you would not have called the PR WIP+EXPERIMENTAL :) So, as I saw something working in the referenced openQA job I consider this great already and assume it's fine? What else do you plan to do before I would be able to approve?

**Comment by Martchus** [comment]
There are already existing tests. For instance `t/29-backend-ipmi.t` checks the `ipmitool` invocation. You could just call `$backend->ipmi_cmdline` again with your new test variable being set and check whether `@ipmi_cmdline` now returns `script …`.

**Comment by czerw** [comment]
>  What else do you plan to do before I would be able to approve?

I need to add unit tests.

**Comment by okurz** [comment]
ok, cool. I would have accepted without but the suggestion from Martchus is good. I would have done the same after we would have merged this PR ;) Give it a try

**Comment by czerw** [comment]
> There are already existing tests. For instance `t/29-backend-ipmi.t` checks the `ipmitool` invocation. You could just call `$backend->ipmi_cmdline` again with your new test variable being set and check whether `@ipmi_cmdline` now returns `script …`.

`$backend->ipmi_cmdline` contains only ipmitool command not  `script` at all. Newly added `script` command is part of `callxterm` function in `localXvnc.pm`, it is used in `eval` section, so it is not part of any return value. Should I somehow mock `eval` function ?

File: backend/ipmi.pm
Comment (czerw) [reply backend/ipmi.pm:100]: I was just following "style" from spvm and hmc backends for this spart, would like to have it same.

File: t/29-backend-ipmi.t
Comment (okurz) [inline t/29-backend-ipmi.t:49]: This fails in CI now with

```
8: 
8: #   Failed test at ./29-backend-ipmi.t line 47.
8: #          got: '0'
8: #     expected: '1'
8: Missing 'xterm-console' at /opt/consoles/localXvnc.pm line 37.
8: # Tests were run but no plan was declared and done_testing() was not seen.
8: # Looks like your test exited with 2 just after 14.
8: [13:31:37] ./29-backend-ipmi.t ...................... 
8: Dubious, test returned 2 (wstat 512, 0x200)
8: Failed 1/14 subtests 
```

because "xterm-console" is not installed in the development environment. I suggest you add the package "xterm-console" in dependencies.yaml after line 91 in the test requirements and call `make update-deps` and submit the changes as well

**Review by okurz — CHANGES_REQUESTED** [review]

File: dependencies.yaml
Comment (okurz) [inline dependencies.yaml:93]: s/Xvnc/xorg-x11-Xvnc/

**Review by Martchus — APPROVED** [review]

**Review by pevik — APPROVED** [review]

File: consoles/localXvnc.pm
Comment (pevik) [reply consoles/localXvnc.pm]: > `script` is part of `util-linux` so we likely don't have to add it as dependency.

@Martchus FYI `script` has been used long time ago in `backend/svirt.pm:272

https://github.com/os-autoinst/os-autoinst/blob/ae94c4bd4c48eb9b8cc0e09b1728c886d0f27c95/backend/svirt.pm#L267-L279

**Pull request merged**
